File: module-1-llm-apis/solutions/01-raw-http.ts
/**
 * Module 1.1: Raw HTTP request to Claude API
 * No SDK — understand the wire format directly.
 */
import "dotenv/config";

const API_KEY = process.env.ANTHROPIC_API_KEY!;
const API_URL = "https://api.anthropic.com/v1/messages";

async function main() {
  const body = {
    model: "claude-sonnet-4-20250514",
    max_tokens: 256,
    messages: [
      { role: "user", content: "What is 2+2? Reply in one sentence." },
    ],
  };

  console.log("→ POST", API_URL);
  console.log("→ Body:", JSON.stringify(body, null, 2));

  const response = await fetch(API_URL, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "x-api-key": API_KEY,
      "anthropic-version": "2023-06-01",
    },
    body: JSON.stringify(body),
  });

  const data = await response.json();
  console.log("\n← Status:", response.status);
  console.log("← Response:", JSON.stringify(data, null, 2));
}

main().catch(console.error);

File: module-1-llm-apis/solutions/python/01_raw_http.py
"""Module 1.1: Raw HTTP request to Claude API."""
import os
import json
import urllib.request
from dotenv import load_dotenv

load_dotenv()

API_KEY = os.environ["ANTHROPIC_API_KEY"]
API_URL = "https://api.anthropic.com/v1/messages"


def main():
    body = {
        "model": "claude-sonnet-4-20250514",
        "max_tokens": 256,
        "messages": [
            {"role": "user", "content": "What is 2+2? Reply in one sentence."},
        ],
    }

    print("→ POST", API_URL)
    print("→ Body:", json.dumps(body, indent=2))

    req = urllib.request.Request(
        API_URL,
        data=json.dumps(body).encode(),
        headers={
            "Content-Type": "application/json",
            "x-api-key": API_KEY,
            "anthropic-version": "2023-06-01",
        },
        method="POST",
    )

    with urllib.request.urlopen(req) as resp:
        data = json.loads(resp.read())
        print(f"\n← Status: {resp.status}")
        print("← Response:", json.dumps(data, indent=2))


if __name__ == "__main__":
    main()
